forge-gui/src/main/java/forge/player/RecordActionsMacroSystem.java

```diff
@@ -0,0 +1,179 @@
+import forge.game.GameEntityView;
+import forge.game.card.CardView;
+import forge.game.player.PlayerView;
+import forge.game.player.actions.FinishTargetingAction;
+import forge.game.player.actions.PassPriorityAction;
+import forge.game.player.actions.PayManaFromPoolAction;
+import forge.game.player.actions.PlayerAction;
+import forge.gamemodes.match.input.*;
+import forge.interfaces.IMacroSystem;
+import forge.util.ITriggerEvent;
+import forge.util.Localizer;
+
+import java.util.List;
+
+// Iteration on the current limited macro system. Instead of asking for IDs to click on
+// Instead we wrap the input queue in a way that we can record what the player is doing and
+// try to play it back as much as possible
+
+public class RecordActionsMacroSystem implements IMacroSystem {
+    private final PlayerControllerHuman playerControllerHuman;
+    private final Localizer localizer = Localizer.getInstance();
+
+    private final List<PlayerAction> actions = Lists.newArrayList();
+    private final List<PlayerAction> playbackActions = Lists.newArrayList();
+
+
+    private boolean recording = false;
+
+    public RecordActionsMacroSystem(PlayerControllerHuman playerControllerHuman) {
+        this.playerControllerHuman = playerControllerHuman;
+    }
+
+    @Override
+    public boolean isRecording() { return recording; }
+
+    @Override
+    public String playbackText() {
+        if (playbackActions.isEmpty()) {
+            return null;
+        }
+
+        return new StringBuilder().append(actions.size() - playbackActions.size()).append(" / ").append(actions.size()).toString();
+    }
+
+    public boolean startRecording() {
+        if (recording) {
+            return false;
+        }
+
+        recording = true;
+        actions.clear();
+        playbackActions.clear();
+        playerControllerHuman.getInputQueue().updateObservers();
+
+        return true;
+    }
+
+    public boolean finishRecording() {
+        if (!recording) {
+            return false;
+        }
+
+        recording = false;
+        playerControllerHuman.getInputQueue().updateObservers();
+
+        return true;
+    }
+
+    @Override
+    public void addRememberedAction(PlayerAction action) {
+        if (!recording) {
+            return;
+        }
+
+        actions.add(action);
+        playbackActions.add(action);
+    }
+
+    @Override
+    public void setRememberedActions() {
+        if (recording) {
+            finishRecording();
+        } else {
+            startRecording();
+        }
+    }
+
+    public void runFullMacro() {
+        if (actions.isEmpty() || recording) {
+            return;
+        }
+
+        if (playbackActions.isEmpty()) {
+            playbackActions.addAll(actions);
+        } else if (actions.size() != playbackActions.size()) {
+            // Not at the beginning of the loop
+            return;
+        }
+
+        while(!playbackActions.isEmpty()) {
+            runFirstAction();
```

Comment:
This isn't hooked up yet, but we need to wait until Input is ready to receive before running the next action.

Reply:
Will look into this in the next one.